Add RFC for extended entity fields (#2598)
elastic/ecs: rfcs/text/0054-extend-entity-fields.md

| entity.attributes.storage_class             | keyword | The storage tier or class assigned to an object storage resource.                                                          |
| entity.attributes.mfa_enabled               | boolean | Indicates whether multi-factor authentication is enabled for this entity.                                                  |
| entity.attributes.granted_permissions       | keyword | The set of permissions explicitly granted to this entity. These are individual action-level permissions, not roles or groups. |
| entity.attributes.known_redirect            | keyword | A known redirect URI or URL associated with this entity, commonly for OAuth applications or services.                      |
| entity.attributes.managed                   | boolean | Indicates whether the entity is managed by an external administration or control system. Applies primarily to host entity. |
| entity.attributes.oauth_consent_restriction | keyword | Indicates any restriction applied to OAuth consent for this entity (e.g., `admin_only`, `verified_only`, `unrestricted`). |

> | entity.lifecycle.last_activity              | date    | The timestamp of the most recent action performed by or attributed to this entity. Distinct from `last_seen`: this implies the entity was active, not just observed. |

Reviewer: we don’t define last_seen don’t we?

Author: Will be calculated by entity store log extraction process, no need to be extracted from integrations data.

| entity.relationships.owns                   | keyword | Captures asset ownership, such as a user owning a host, email address, Windows SID, or employee ID.                       |
| entity.relationships.depends_on             | keyword | Identifiers of entities that this entity depends on to function.                                                           |
| entity.relationships.supervises             | keyword | Identifiers of entities that this entity supervises, manages, or is responsible for.                                       |
| entity.relationships.administrators         | keyword | The list of identities (usernames, emails, or IDs) that administer this entity.                                            |

> This RFC also proposes enabling `entity.risk.*` field reuse so the existing ECS `risk.*` fields can be nested under `entity` when the risk score describes the normalized entity rather than a specific entity-type field set.

Author: Yes, removed from this one.
Will create a dedicated one for `entity.risk`.

The `entity.relationships.*` leaves are meant to support graph enrichment, dependency mapping, ownership modeling, and organizational context. Using concrete relationship leaves instead of an open-ended `entity.relationship` object should make the schema easier for ECS users to understand, query, and review. Enabling `entity.risk.*` under `entity` allows risk scores to be attached to entities generically, rather than requiring different placement depending on whether the normalized entity is a host, user, service, or a more generic type.

Example:

```json
{
  "@timestamp": "2026-03-31T12:00:00Z",
  "user": {
    "name": "paul",
    "entity": {
      "type": "user",
      "attributes": {
        "mfa_enabled": true,
        "granted_permissions": ["read", "write", "reset_password"]
      },
      "lifecycle": {
        "last_activity": "2026-03-31T11:42:00Z"
      },
      "risk": {
        "calculated_level": "Moderate",
        "calculated_score": 47.3,
        "calculated_score_norm": 47.3
      },
      "relationships": {

>         "owns": ["host:workstation-123"],

Reviewer: I agree with the object here, but the object should be defined.

>         "owns": ["host:workstation-123"],

Author: Same as here: https://github.com/elastic/ecs/pull/2598#discussion_r3073860027

>         "supervises": ["user:erik@okta"]

Author: You’re right. The catch is that ESQL (our solution is based on it) doesn’t support [nested mappings](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/nested), so we can’t persist that structure in the entity store indices.
What we can do is instruct users and integrations to emit this structure, and in our entity-building queries apply the prioritization we already use today: prefer email, then id, then name.
cc @romulets

>         "supervises": ["user:erik@okta"]

Author: I discussed it with @romulets, ESQL limited nested support is blocking us from querying a list of objects.
But for `entity.relationships.*` you can publish a single object with a bunch of keyword array fields: `entity_ids`, `host_ids`, `user_ids`, `emails`, `hostnames`, `usernames`, etc.
So we stay flat and don’t need nested for ESQL.
Downside is we’re not pairing “this email goes with this id”, it’s more of a bag of identifiers we can match on. The Upside is that integrations can throw in whatever identifiers they have and we can still extend the list over time.

Example:
```
{
  "entity_id": ["entity:abc123"],
  "host_id": ["host-uuid-123", "host-uuid-456"],
  "user_id": ["00u123"],
  "email": ["supervisor@corp.example"],
  "hostname": [],
  "username": ["jsmith"]
}
```

>         "supervises": ["user:erik@okta"]

Reviewer 2: I don't have a problem from the integration side to deliver the data that way if that's the only way you can query later. It should be well documented though so developers of new integrations and people that populate these fields in the future are aware of this.

@andrewkroh what do you think about this way of storing the data without pairing? is this already happening for any other fields in ECS?

>         "supervises": ["user:erik@okta"]

Reviewer: why would you do hostname and host_id separately? why not using host.hostname and host.id etc?

>         "supervises": ["user:erik@okta"]

Author: No reason, will update.

>         "supervises": ["user:erik@okta"]

Reviewer 2: Another question I have with `relationships` fields, are they going to store predefined fields inside? such as `id`, `email`, `username`?

In case they don't, we might have the risk to have same data with different names in different integrations (e.g., username and name both storing an entity username). In addition, in case their content is undefined, should them be categorised as flattened?

>         "supervises": ["user:erik@okta"]

Reviewer: The goal of this PR is actually to define all fields so there will be no `object` anymore, but detailed data structure. this should resolve problem for different integrations to understand the schema.

>         "supervises": ["user:erik@okta"]

Author: Yes the object will consist of the following identifiers:
```
{
  "entity.id": ["entity:abc123"],
  "host.id": ["host-uuid-123", "host-uuid-456"],
  "user.id": ["00u123"],
  "email": ["supervisor@corp.example"],
  "host.name": [],
  "user.name": ["jsmith"]
}
```

      }
    }
  }
}
```

## Source data

Potential sources of data include:

- identity provider and SaaS audit logs
- cloud asset discovery logs
- endpoint and device management platforms
- other integrations that already populate or derive normalized entity metadata

## Scope of impact

At a high level, this proposal should be additive. Producers that already emit `host.entity.*`, `user.entity.*`, `service.entity.*`, or root `entity.*` would be able to populate these leaves directly. Consumers such as entity stores, graph enrichment systems, and investigation workflows would gain a more concrete and queryable schema for ownership, dependencies, supervision, access posture, and entity-level risk.

## Concerns

> - Should `entity.relationships.administrators` remain a multi-valued `keyword` list of identifiers, or should ECS eventually define a richer identity structure for administrator references?

Reviewer: You can define it yourself in this proposal if it’s something you need. But I don’t think we need to define administrators as separate namespace if that’s what you asked for

- Should `entity.attributes.known_redirect` remain singular in name while allowing multiple values, or should ECS prefer a pluralized field name?

> - Once `entity.risk.*` is enabled under `entity`, should ECS explicitly recommend avoiding duplicate risk storage under entity-type-specific paths when the same score describes the normalized entity?

Reviewer: `risk.*` fields are not a part of this PR anynore

| entity.attributes.known_redirects | keyword (list) | Service | Known redirect URIs or URLs associated with this entity, commonly for OAuth applications or services. Plural name reflects ECS guidance to use singular/plural names to reflect field content. |
| entity.attributes.managed | boolean | Host, Service | Indicates whether the entity is managed by an external administration or control system. |
| entity.attributes.oauth_consent_restriction | keyword | Service | Restriction applied to OAuth consent for this entity (e.g. `admin_only`, `verified_only`, `unrestricted`). **Values:** integration-defined keywords unless standardized later. |
| entity.lifecycle.last_activity | date | User, Host, Service | Timestamp of the most recent action performed by or attributed to this entity (active use). Distinct from **`entity.last_seen_timestamp`**, which records when the entity was last observed in data; `last_activity` implies the entity was active, not only seen. |

> | entity.relationships.owns | object | all | Identifiers of assets or identities this entity owns. Structure is a **flat bag of keyword arrays** (see [Relationship identifier structure](#relationship-identifier-structure)), not an array of nested objects, so storage and querying remain compatible with ESQL limitations on nested mappings. |

Reviewer: below there is just an example of such event. Is this example binding? If so it should be defined in this PR accordingly with names, types, examples etc. This applies to every object in this table

Also we have updated and streamlined the RFC process
https://github.com/elastic/ecs/tree/main/rfcs

Currently there is only 1 stage and requirement to create also yaml file with fields if RFC describes them

Author: Yes, it's binding as this this the structure we're expecting in the logs extraction process.
Hopes it's better defined now.
I saw the RFC process has been updated, do we expect this RFC to be updated accordingly?

| entity.attributes.known_redirects | keyword (list) | Service | Known redirect URIs or URLs associated with this entity, commonly for OAuth applications or services. Plural name reflects ECS guidance to use singular/plural names to reflect field content. |
| entity.attributes.managed | boolean | Host, Service | Indicates whether the entity is managed by an external administration or control system. |
| entity.attributes.oauth_consent_restriction | keyword | Service | Restriction applied to OAuth consent for this entity (e.g. `admin_only`, `verified_only`, `unrestricted`). **Values:** integration-defined keywords unless standardized later. |
| entity.lifecycle.last_activity | date | User, Host, Service | Timestamp of the most recent action performed by or attributed to this entity (active use). Distinct from **`entity.last_seen_timestamp`**, which records when the entity was last observed in data; `last_activity` implies the entity was active, not only seen. |

> | entity.relationships.owns | object | all | Identifiers of assets or identities this entity owns. Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). Same rules apply to every `entity.relationships.*` in this table. |

Reviewer: ```suggestion
| entity.relationships.owns | object | all | Identifiers of assets or identities this entity owns. Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
```

> | entity.relationships.depends_on | object | all | Identifiers of entities this entity **requires for operation**—for example: a service depending on a database or upstream API, an application depending on an identity provider, or a workload depending on a host or cluster. Same object shape and **allowed keys** as `owns`. |

Reviewer: ```suggestion
| entity.relationships.depends_on | object | all | Identifiers of entities this entity **requires for operation**—for example: a service depending on a database or upstream API, an application depending on an identity provider, or a workload depending on a host or cluster. Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
```

> | entity.relationships.supervises | object | all | Identifiers of entities this entity supervises, manages, or is responsible for (e.g. manager–reporting-line or org hierarchy). Same object shape and **allowed keys** as `owns`. |

Reviewer: ```suggestion
| entity.relationships.supervises | object | all | Identifiers of entities this entity supervises, manages, or is responsible for (e.g. manager–reporting-line or org hierarchy). Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
```

> | entity.relationships.administered_by | object | all | Identifiers of identities that administer this entity (incoming relationship: “who administers this entity”). Same object shape and **allowed keys** as `owns`. |

Reviewer: ```suggestion
| entity.relationships.administered_by | object | all | Identifiers of identities that administer this entity (incoming relationship: “who administers this entity”). Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
```

> _Indicative only; specific integrations may justify exceptions. **all** under Applied Entity Type means User, Host, or Service whenever the relationship applies to that normalized entity._

Reviewer: as of now we don’t have a built-in option to specify that some of the parameters applies only for specific entity types. This should be added verbatim to every field as additional note

**Applied Entity Type** lists which normalized entity kinds (**User**, **Host**, **Service**) the field is typically relevant for. Comma-separated means any of those; **all** means User, Host, and Service when semantically appropriate.

| Field | Type | Applied Entity Type | Description |
| ----- | ---- | ------------------- | ----------- |
| entity.attributes.storage_class | keyword | Service | The storage tier or class assigned to an object storage resource (e.g. S3/GCS/Azure object tiers). **Values:** free-text or provider-specific labels unless/until ECS defines `allowed_values`; common examples include `STANDARD`, `STANDARD_IA`, `GLACIER`, `COLDLINE` (not an exhaustive enum in this RFC). |
| entity.attributes.mfa_enabled | boolean | User | Indicates whether multi-factor authentication is enabled for this entity. |
| entity.attributes.permissions | keyword (list) | User, Host, Service | Action-level permissions associated with this entity (not roles or groups). Prefer this name over `granted_permissions` unless a future RFC introduces parallel fields for inherited or effective permissions. |

> | entity.attributes.known_redirects | keyword (list) | Service | Known redirect URIs or URLs associated with this entity, commonly for OAuth applications or services. Plural name reflects ECS guidance to use singular/plural names to reflect field content. |

Reviewer: ```suggestion
| entity.attributes.known_redirects | keyword (list) | Service | Known redirect URIs or URLs associated with this entity, commonly for OAuth applications or services. |
```

why do we need this sentence?

Each relationship object holds parallel identifier arrays keyed only by the allowed property names; it does not preserve which `user.email` value pairs with which `user.id` in a single structure. Correlation and pairing are expected to be resolved by entity-building logic or downstream normalization (similar in spirit to populating `related.user`, `related.host`, etc. with parallel arrays).

### Allowed keys on `entity.relationships.*` objects

Values in each array must follow the same conventions as the corresponding ECS field (e.g. `user.email` values are email-shaped strings; `user.domain` values follow ECS directory or AD/LDAP domain name semantics; `host.id` values follow host id semantics from ECS).

| Property key (JSON) | ECS type | `normalize: [array]` | Semantics |
| ------------------- | -------- | --------------------- | ----------- |
| `entity.id` | keyword | yes | Identifiers of referenced entities, using the same meaning as root `entity.id` (stable id for correlation within scope). |
| `host.id` | keyword | yes | Referenced host ids. |

> | `host.name` | keyword | yes | Referenced host names. |

Reviewer 2: Does it make sense to add here host.domain as well?

Author: We can add, for the user broader context, but we don't utilize `host.domain` in the entity store.
Identity ranking for hosts is `host.id` → `host.name` → `host.hostname` only.
`host.domain` is not in that ranking.

| entity.relationships.depends_on | object | all | Identifiers of entities this entity **requires for operation**—for example: a service depending on a database or upstream API, an application depending on an identity provider, or a workload depending on a host or cluster. Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
| entity.relationships.supervises | object | all | Identifiers of entities this entity supervises, manages, or is responsible for (e.g. manager–reporting-line or org hierarchy). Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |
| entity.relationships.administers | object | all | Identifiers of entities that this entity administers (for example managed hosts, delegated accounts, or administered resources). Value **must** be an object whose keys are those defined in [Allowed keys on `entity.relationships.*` objects](#allowed-keys-on-entityrelationships-objects). |

> 

Reviewer: I don’t quite understand how this differs from `supervises`?

Author: **Supervises** applies to user entities and captures management hierarchy: who reports to whom as people - “who is whose manager in the org?” It’s about org chain of command, not technical privileges.

**Administers** means this user is an administrator of something technical, for example Entra ID, Microsoft 365, or a GCP organization, not merely “uses” those systems with normal permissions.

@erikh-elastic — does this match your understanding, or should we adjust anything?

Reviewer: from my pov it’s very subtle difference and it’s not clear from description. Does it mean first is only applicable to user entities and second to all? i can’t tell the difference from the field name and description doesn’t help much. After your explanation it’s a bit better but still not 100% clear. Can we improve it?

Reviewer 2: I just want to highlight that changing the direction of the field (from `identities that administer this entity` to `entities that this entity administers`) at this point, it also affects the changes that we already made to the integrations to support this field.

They have not been merged yet, but it will require extra effort to support a new field, that could also lead to adding changes to the Filebeat input and wait for a new release to get it working.

Reviewer 3: Yes, thank you. @uri-weisman! That is exactly it.

From a security product perspective "Supervises" versus "Administrator" are completely different things.

Supervises is a manager to employee type relationship versus administrator is a user to a host or service or etc type relationship.

Author: @trisch-me
I’ll tighten up the description and will add examples that highlight supervises as a user→user relationship and administers is user→service/host, but I’d rather keep the description a bit open-ended regarding types so we’re not tied only to those.

@chemamartinez Totally agree, and thanks for surfacing it. We want this data to be valuable for customers. Thinking it through again, the passive direction (“identities that administer this entity”) makes less sense, so If reversing the direction turns into a heavy lift, we’re good deferring that to the next iteration.
